File: lib/go/node/uncompressed.go
package node

import (
	"net"
	"sync/atomic"
	"unsafe"

	types "github.com/Mwangi-Derrick/radixip/lib/go"
)

// atomicNode implements RadixNode using atomic operations for thread safety
type atomicNode struct {
	bit      unsafe.Pointer // *uint8
	left     unsafe.Pointer // *atomicNode
	right    unsafe.Pointer // *atomicNode
	metadata unsafe.Pointer // *types.Metadata
	prefix   unsafe.Pointer // *net.IPNet
}

type normalNode struct {
	bit      uint8
	left     *normalNode
	right    *normalNode
	metadata *types.Metadata
	prefix   *net.IPNet
}

type paddedNode struct {
	_pad1    [64]byte
	bit      uint8
	_pad2    [56]byte
	left     *paddedNode
	_pad3    [56]byte
	right    *paddedNode
	_pad4    [56]byte
	metadata *types.Metadata
	_pad5    [56]byte
	prefix   *net.IPNet
	_pad6    [56]byte
}

func NewAtomicNode() *atomicNode {
	return &atomicNode{}
}

func (n *atomicNode) Bit() *uint8 {
	return (*uint8)(atomic.LoadPointer(&n.bit))
}

func (n *atomicNode) SetBit(bit uint8) {
	bitVal := bit
	atomic.StorePointer(&n.bit, unsafe.Pointer(&bitVal))
}

func (n *atomicNode) Left() types.RadixNode {
	ptr := atomic.LoadPointer(&n.left)
	if ptr == nil {
		return nil
	}
	return (*atomicNode)(ptr)
}

func (n *atomicNode) SetLeft(node types.RadixNode) {
	if node == nil {
		atomic.StorePointer(&n.left, nil)
	} else {
		atomic.StorePointer(&n.left, unsafe.Pointer(node.(*atomicNode)))
	}
}

func (n *atomicNode) Right() types.RadixNode {
	ptr := atomic.LoadPointer(&n.right)
	if ptr == nil {
		return nil
	}
	return (*atomicNode)(ptr)
}

func (n *atomicNode) SetRight(node types.RadixNode) {
	if node == nil {
		atomic.StorePointer(&n.right, nil)
	} else {
		atomic.StorePointer(&n.right, unsafe.Pointer(node.(*atomicNode)))
	}
}

func (n *atomicNode) Metadata() *types.Metadata {
	return (*types.Metadata)(atomic.LoadPointer(&n.metadata))
}

func (n *atomicNode) SetMetadata(metadata *types.Metadata) {
	atomic.StorePointer(&n.metadata, unsafe.Pointer(metadata))
}

func (n *atomicNode) ClearMetadata() {
	atomic.StorePointer(&n.metadata, nil)
}

func (n *atomicNode) Prefix() *net.IPNet {
	return (*net.IPNet)(atomic.LoadPointer(&n.prefix))
}

func (n *atomicNode) SetPrefix(prefix *net.IPNet) {
	atomic.StorePointer(&n.prefix, unsafe.Pointer(prefix))
}

func NewNormalNode() *normalNode {
	return &normalNode{}
}

func (n *normalNode) Bit() *uint8 {
	return &n.bit
}

func (n *normalNode) SetBit(bit uint8) {
	n.bit = bit
}

func (n *normalNode) Left() types.RadixNode {
	return n.left
}

func (n *normalNode) SetLeft(node types.RadixNode) {
	n.left = node.(*normalNode)
}

func (n *normalNode) Right() types.RadixNode {
	return n.right
}

func (n *normalNode) SetRight(node types.RadixNode) {
	n.right = node.(*normalNode)
}

func (n *normalNode) Metadata() *types.Metadata {
	return n.metadata
}

func (n *normalNode) SetMetadata(metadata *types.Metadata) {
	n.metadata = metadata
}

func (n *normalNode) ClearMetadata() {
	n.metadata = nil
}

func (n *normalNode) Prefix() *net.IPNet {
	return n.prefix
}

func (n *normalNode) SetPrefix(prefix *net.IPNet) {
	n.prefix = prefix
}

func NewPaddedNode() *paddedNode {
	return &paddedNode{}
}

func (n *paddedNode) Bit() *uint8 {
	return &n.bit
}

func (n *paddedNode) SetBit(bit uint8) {
	n.bit = bit
}

func (n *paddedNode) Left() types.RadixNode {
	return n.left
}

func (n *paddedNode) SetLeft(node types.RadixNode) {
	n.left = node.(*paddedNode)
}

func (n *paddedNode) Right() types.RadixNode {
	return n.right
}

func (n *paddedNode) SetRight(node types.RadixNode) {
	n.right = node.(*paddedNode)
}

func (n *paddedNode) Metadata() *types.Metadata {
	return n.metadata
}

func (n *paddedNode) SetMetadata(metadata *types.Metadata) {
	n.metadata = metadata
}

func (n *paddedNode) ClearMetadata() {
	n.metadata = nil
}

func (n *paddedNode) Prefix() *net.IPNet {
	return n.prefix
}

func (n *paddedNode) SetPrefix(prefix *net.IPNet) {
	n.prefix = prefix
}

File: lib/rust/src/node/uncompressed.rs
//! Uncompressed (bit-by-bit) radix trie node implementations.
//!
//! These nodes store a single `bit` value and left/right children,
//! traversed one bit at a time. There are three concurrency variants:
//!
//! - [`NormalNode`]  — RwLock on every field  
//! - [`AtomicNode`]  — AtomicU8 for the bit, RwLock for children  
//! - [`PaddedNode`]  — Cache-line padded to avoid false sharing  
//! - [`LockFreeNode`] — DashMap-based children for fully lock-free reads  

use dashmap::DashMap;
use ipnetwork::IpNetwork;
use std::sync::{
    Arc, RwLock,
    atomic::{AtomicU8, Ordering},
};

use crate::traits::RadixNode;
use crate::types::Metadata;

//
// NORMAL NODE (RwLock on all fields)
//

#[derive(Default)]
pub struct NormalNode {
    bit: RwLock<Option<u8>>,
    left: RwLock<Option<Arc<dyn RadixNode>>>,
    right: RwLock<Option<Arc<dyn RadixNode>>>,
    metadata: RwLock<Option<Metadata>>,
    prefix: RwLock<Option<IpNetwork>>,
}

impl NormalNode {
    pub fn new() -> Self {
        Self::default()
    }
}

impl RadixNode for NormalNode {
    fn bit(&self) -> Option<u8> {
        *self.bit.read().unwrap()
    }

    fn left(&self) -> Option<Arc<dyn RadixNode>> {
        self.left.read().unwrap().clone()
    }

    fn right(&self) -> Option<Arc<dyn RadixNode>> {
        self.right.read().unwrap().clone()
    }

    fn metadata(&self) -> Option<Metadata> {
        self.metadata.read().unwrap().clone()
    }

    fn prefix(&self) -> Option<IpNetwork> {
        *self.prefix.read().unwrap()
    }

    fn set_left(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.left.write().unwrap() = node;
    }

    fn set_right(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.right.write().unwrap() = node;
    }

    fn set_metadata(&self, metadata: Metadata) {
        *self.metadata.write().unwrap() = Some(metadata);
    }

    fn clear_metadata(&self) {
        *self.metadata.write().unwrap() = None;
    }

    fn set_bit(&self, bit: u8) {
        *self.bit.write().unwrap() = Some(bit);
    }

    fn set_prefix(&self, prefix: IpNetwork) {
        *self.prefix.write().unwrap() = Some(prefix);
    }
}

//
// ATOMIC NODE (AtomicU8 bit + RwLock children)
//

#[repr(C)]
pub struct AtomicNode {
    /// Encoded as 0 = None, n+1 = Some(n) so we can use AtomicU8.
    bit: AtomicU8,
    left: RwLock<Option<Arc<dyn RadixNode>>>,
    right: RwLock<Option<Arc<dyn RadixNode>>>,
    metadata: RwLock<Option<Metadata>>,
    prefix: RwLock<Option<IpNetwork>>,
}

impl AtomicNode {
    pub fn new() -> Self {
        Self {
            bit: AtomicU8::new(0),
            left: RwLock::new(None),
            right: RwLock::new(None),
            metadata: RwLock::new(None),
            prefix: RwLock::new(None),
        }
    }

    pub fn with_bit(bit: u8) -> Self {
        let node = Self::new();
        node.set_bit(bit);
        node
    }
}

impl Default for AtomicNode {
    fn default() -> Self {
        Self::new()
    }
}

impl RadixNode for AtomicNode {
    fn bit(&self) -> Option<u8> {
        let b = self.bit.load(Ordering::Acquire);
        if b == 0 { None } else { Some(b - 1) }
    }

    fn left(&self) -> Option<Arc<dyn RadixNode>> {
        self.left.read().unwrap().clone()
    }

    fn right(&self) -> Option<Arc<dyn RadixNode>> {
        self.right.read().unwrap().clone()
    }

    fn metadata(&self) -> Option<Metadata> {
        self.metadata.read().unwrap().clone()
    }

    fn prefix(&self) -> Option<IpNetwork> {
        *self.prefix.read().unwrap()
    }

    fn set_left(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.left.write().unwrap() = node;
    }

    fn set_right(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.right.write().unwrap() = node;
    }

    fn set_metadata(&self, metadata: Metadata) {
        *self.metadata.write().unwrap() = Some(metadata);
    }

    fn clear_metadata(&self) {
        *self.metadata.write().unwrap() = None;
    }

    fn set_bit(&self, bit: u8) {
        // Store bit+1 so 0 encodes None
        self.bit.store(bit + 1, Ordering::Release);
    }

    fn set_prefix(&self, prefix: IpNetwork) {
        *self.prefix.write().unwrap() = Some(prefix);
    }
}

//
// PADDED NODE (Cache-line aligned, 64-byte padding between fields)
//

#[repr(C, align(64))]
pub struct PaddedNode {
    _pad1: [u8; 64],
    bit: RwLock<Option<u8>>,
    _pad2: [u8; 56],
    left: RwLock<Option<Arc<dyn RadixNode>>>,
    _pad3: [u8; 56],
    right: RwLock<Option<Arc<dyn RadixNode>>>,
    _pad4: [u8; 56],
    metadata: RwLock<Option<Metadata>>,
    _pad5: [u8; 56],
    prefix: RwLock<Option<IpNetwork>>,
    _pad6: [u8; 56],
}

impl PaddedNode {
    pub fn new() -> Self {
        Self {
            bit: RwLock::new(None),
            left: RwLock::new(None),
            right: RwLock::new(None),
            metadata: RwLock::new(None),
            prefix: RwLock::new(None),
            _pad1: [0; 64],
            _pad2: [0; 56],
            _pad3: [0; 56],
            _pad4: [0; 56],
            _pad5: [0; 56],
            _pad6: [0; 56],
        }
    }
}

impl Default for PaddedNode {
    fn default() -> Self {
        Self::new()
    }
}

impl RadixNode for PaddedNode {
    fn bit(&self) -> Option<u8> {
        *self.bit.read().unwrap()
    }

    fn left(&self) -> Option<Arc<dyn RadixNode>> {
        self.left.read().unwrap().clone()
    }

    fn right(&self) -> Option<Arc<dyn RadixNode>> {
        self.right.read().unwrap().clone()
    }

    fn metadata(&self) -> Option<Metadata> {
        self.metadata.read().unwrap().clone()
    }

    fn prefix(&self) -> Option<IpNetwork> {
        *self.prefix.read().unwrap()
    }

    fn set_left(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.left.write().unwrap() = node;
    }

    fn set_right(&self, node: Option<Arc<dyn RadixNode>>) {
        *self.right.write().unwrap() = node;
    }

    fn set_metadata(&self, metadata: Metadata) {
        *self.metadata.write().unwrap() = Some(metadata);
    }

    fn clear_metadata(&self) {
        *self.metadata.write().unwrap() = None;
    }

    fn set_bit(&self, bit: u8) {
        *self.bit.write().unwrap() = Some(bit);
    }

    fn set_prefix(&self, prefix: IpNetwork) {
        *self.prefix.write().unwrap() = Some(prefix);
    }
}

//
// LOCK-FREE NODE (DashMap-based, for maximum concurrent read throughput)
//

/// Key enum used inside the DashMap for left/right children.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum ChildKey {
    Left,
    Right,
}

pub struct LockFreeNode {
    bit: AtomicU8,
    children: DashMap<ChildKey, Arc<dyn RadixNode>>,
    metadata: RwLock<Option<Metadata>>,
    prefix: RwLock<Option<IpNetwork>>,
}

impl LockFreeNode {
    pub fn new() -> Self {
        Self {
            bit: AtomicU8::new(0),
            children: DashMap::new(),
            metadata: RwLock::new(None),
            prefix: RwLock::new(None),
        }
    }
}

impl Default for LockFreeNode {
    fn default() -> Self {
        Self::new()
    }
}

impl RadixNode for LockFreeNode {
    fn bit(&self) -> Option<u8> {
        let b = self.bit.load(Ordering::Acquire);
        if b == 0 { None } else { Some(b - 1) }
    }

    fn left(&self) -> Option<Arc<dyn RadixNode>> {
        self.children
            .get(&ChildKey::Left)
            .map(|r| r.value().clone())
    }

    fn right(&self) -> Option<Arc<dyn RadixNode>> {
        self.children
            .get(&ChildKey::Right)
            .map(|r| r.value().clone())
    }

    fn metadata(&self) -> Option<Metadata> {
        self.metadata.read().unwrap().clone()
    }

    fn prefix(&self) -> Option<IpNetwork> {
        *self.prefix.read().unwrap()
    }

    fn set_left(&self, node: Option<Arc<dyn RadixNode>>) {
        match node {
            Some(n) => {
                self.children.insert(ChildKey::Left, n);
            }
            None => {
                self.children.remove(&ChildKey::Left);
            }
        }
    }

    fn set_right(&self, node: Option<Arc<dyn RadixNode>>) {
        match node {
            Some(n) => {
                self.children.insert(ChildKey::Right, n);
            }
            None => {
                self.children.remove(&ChildKey::Right);
            }
        }
    }

    fn set_metadata(&self, metadata: Metadata) {
        *self.metadata.write().unwrap() = Some(metadata);
    }

    fn clear_metadata(&self) {
        *self.metadata.write().unwrap() = None;
    }

    fn set_bit(&self, bit: u8) {
        self.bit.store(bit + 1, Ordering::Release);
    }

    fn set_prefix(&self, prefix: IpNetwork) {
        *self.prefix.write().unwrap() = Some(prefix);
    }
}
